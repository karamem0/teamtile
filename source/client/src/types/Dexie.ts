//
// Copyright (c) 2023 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

export interface ArrayEntity<T> {
  id: string,
  expired: number,
  values: T[]
}

export interface ValueEntity<T> {
  id: string,
  expired: number,
  value: T
}
