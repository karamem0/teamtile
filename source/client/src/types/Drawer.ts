//
// Copyright (c) 2021-2025 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

export type DrawerType =
  | 'channel'
  | 'member'
  | 'tag';
