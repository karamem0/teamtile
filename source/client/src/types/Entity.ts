//
// Copyright (c) 2021-2024 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

export enum MembershipType {
  private = 'private',
  standard = 'standard'
}

export enum VisibilityType {
  private = 'private',
  public = 'public'
}

export enum AccordionType {
  all = 'all',
  archived = 'archived',
  pinned = 'pinned'
}

export interface Channel {
  id?: string,
  displayName?: string,
  membershipType?: MembershipType,
  webUrl?: string
}

export interface Drive {
  id?: string,
  webUrl?: string
}

export interface Group {
  id?: string,
  email?: string,
  sensitivityLabel?: string
}

export interface Icon {
  id?: string,
  data?: string
}

export interface Member {
  id?: string,
  displayName?: string,
  email?: string,
  icon?: string,
  userId?: string
}

export interface Tab {
  id?: string,
  appId?: string,
  displayName?: string,
  webUrl?: string
}

export interface Team {
  id?: string,
  archived?: boolean,
  description?: string,
  displayName?: string,
  icon?: string,
  internalId?: string,
  visibility?: VisibilityType,
  webUrl?: string
}

export interface TeamInfo {
  id?: string,
  displayName?: string,
  tenantId?: string
}
