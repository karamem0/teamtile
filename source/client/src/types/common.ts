//
// Copyright (c) 2021 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/master/LICENSE
//

export interface KeyValue<K, V> {
  key: K,
  value: V
}
