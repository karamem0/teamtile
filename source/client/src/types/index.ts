//
// Copyright (c) 2021 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/master/LICENSE
//

export interface Team {
  id?: string;
  internalId?: string;
  name?: string;
  description?: string;
  visibility?: string;
  url?: string;
  icon?: string;
}

export interface TeamChannels {
  count?: number;
  nextLink?: string;
  values?: TeamChannel[];
}

export interface TeamChannel {
  id?: string;
  name?: string;
  url?: string;
}

export interface TeamMembers {
  count?: number;
  nextLink?: string;
  values?: TeamMember[];
}

export interface TeamMember {
  id?: string;
  name?: string;
  email?: string;
}

export interface TeamDrive {
  id?: string;
  url?: string;
}
