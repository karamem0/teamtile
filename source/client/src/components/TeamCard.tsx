import React from 'react';
import * as microsoftTeams from '@microsoft/teams-js';
import {
  Button,
  Card,
  Text
} from '@fluentui/react-northstar';
import { LockIcon, GlobeIcon } from '@fluentui/react-icons';
import { Team } from '../types';
import AppContext from '../contexts/AppContext';
import useTeam from '../hooks/useTeam';
import TeamIcon from './TeamIcon';
import TeamMenu from './TeamMenu';

interface TeamCardProps {
  team?: Team;
}

const TeamCard: React.FC<TeamCardProps> = (props: TeamCardProps) => {

  const [ token, , , setError ] = React.useContext(AppContext);
  const [
    team,
    channels,
    members,
    drive,
    error
  ] = useTeam({ token, team: props.team });

  React.useEffect(() => {
    if (!setError) {
      return;
    }
    setError(error);
  }, [ setError, error ]);

  return (
    <React.Fragment>
      {
        team &&
          <Card
            className="card"
            fluid>
            <div className="card-column">
              <div className="card-column-item">
                <TeamIcon
                  icon={team.icon}
                  name={team.name} />
              </div>
              <div className="card-column-item">
                <div className="card-row">
                  <Button
                    className="card-name"
                    content={team.name}
                    text
                    onClick={() => {
                      if (!team.url) {
                        return;
                      }
                      microsoftTeams.executeDeepLink(team.url);
                    }} />
                  <Text
                    className="card-description"
                    content={team.description}
                    size="small"
                    truncated />
                  <TeamMenu
                    channels={channels}
                    drive={drive}
                    members={members}
                    team={team} />
                </div>
              </div>
              <div className="card-column-item">
                <Text color="brand">
                  {
                    team.visibility === 'public' &&
                      <GlobeIcon className="card-menu-icon" />
                  }
                  {
                    team.visibility === 'private' &&
                      <LockIcon className="card-menu-icon" />
                  }
                </Text>
              </div>
            </div>
          </Card>
      }
    </React.Fragment>
  );

};

export default TeamCard;
