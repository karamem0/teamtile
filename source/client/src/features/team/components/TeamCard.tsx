//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { app } from '@microsoft/teams-js';

import { Item } from '../../../types/Store';

import Presenter from './TeamCard.presenter';

export interface TeamCardProps {
  item: Item
}

function TeamCard(props: TeamCardProps) {

  const { item } = props;

  const handleClick = React.useCallback(() => {
    if (!item.value.webUrl) {
      return;
    }
    app.openLink(item.value.webUrl);
  }, [ item ]);

  return (
    <Presenter
      item={item}
      onClick={handleClick} />
  );

}

export default TeamCard;
