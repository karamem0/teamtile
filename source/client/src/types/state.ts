//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import { KeyValue } from './common';
import {
  Channel,
  Drive,
  Group,
  MemberWithIcon,
  TeamWithIcon
} from './entity';

export type Item = (
  KeyValue<ItemKey, ItemValue | null> &
  {
    loading: boolean | undefined,
    visible: boolean | undefined
  }
)

export type ItemKey = string;

export type ItemValue = (
  TeamWithIcon &
  Group &
  {
    channels: Channel[] | undefined,
    drive: Drive | undefined,
    members: MemberWithIcon[] | undefined
  }
);

export interface State {
  itemFilter: string | null | undefined,
  items: Item[],
  loading: boolean | undefined
}
