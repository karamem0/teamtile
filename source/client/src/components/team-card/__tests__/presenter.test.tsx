//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

jest.mock('../../channel-menu-item', () => ({
  ChannelMenuItem: ({ children }: React.PropsWithChildren<unknown>) => (
    <div data-testid="ChannelMenuItem">
      {children}
    </div>
  )
}));

jest.mock('../../drive-menu-item', () => ({
  DriveMenuItem: ({ children }: React.PropsWithChildren<unknown>) => (
    <div data-testid="DriveMenuItem">
      {children}
    </div>
  )
}));

jest.mock('../../member-menu-item', () => ({
  MemberMenuItem: ({ children }: React.PropsWithChildren<unknown>) => (
    <div data-testid="MemberMenuItem">
      {children}
    </div>
  )
}));

import React from 'react';

import { render, screen } from '@testing-library/react';

import { Item } from '../../../types/state';
import TeamCard from '../presenter';

beforeEach(() => {
  jest.clearAllMocks();
  jest.restoreAllMocks();
});

describe('TeamCard', () => {

  it('create shapshot', async () => {
    const json = await import('./__jsons__/presenter.test.json');
    const params = {
      item: json.default as Item
    };
    render(<TeamCard {...params} />);
    expect(screen.queryAllByText(/^.*$/)[0]).toMatchSnapshot();
  });

});
