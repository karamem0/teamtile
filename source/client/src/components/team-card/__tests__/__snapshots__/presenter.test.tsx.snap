// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamCard create shapshot when item is loaded 1`] = `
.emotion-0 {
  height: 6rem;
}

.emotion-1 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: auto 1fr;
  gap: 0.5rem;
}

.emotion-2 {
  display: grid;
  gap: 0.25rem;
}

.emotion-3 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: 1fr auto auto;
  gap: 0.25rem;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.emotion-4 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.emotion-5 {
  cursor: pointer;
}

.emotion-6 {
  width: 1rem;
  height: 1rem;
  vertical-align: top;
}

.emotion-8 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: auto auto auto auto;
  gap: 1rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: left;
  -ms-flex-pack: left;
  -webkit-justify-content: left;
  justify-content: left;
}

<body>
  <div>
    <div
      class="ui-card emotion-0"
      data-aa-class="Card"
      data-uses-unhanded-props="true"
      role="listitem"
    >
      <div
        class="emotion-1"
      >
        <div
          class="ui-avatar"
          data-uses-unhanded-props="true"
        >
          <span
            class="ui-label"
            data-uses-unhanded-props="true"
            title="HR Taskforce"
          >
            HT
          </span>
        </div>
        <div
          class="emotion-2"
        >
          <div
            class="emotion-3"
          >
            <div
              class="emotion-4"
            >
              <span
                class="ui-text emotion-5"
                data-uses-unhanded-props="true"
                dir="auto"
                role="button"
              >
                HR Taskforce
              </span>
            </div>
            <span
              class="ui-text"
              data-uses-unhanded-props="true"
            />
            <span
              class="ui-text"
              data-uses-unhanded-props="true"
            >
              <span
                aria-hidden="true"
                class="root_dd790ee3 emotion-6"
              >
                <svg
                  class="svg_dd790ee3"
                  focusable="false"
                  viewBox="0 0 2048 2048"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1792 896v1152H256V896h256V522q0-108 39-203t108-166T821 41t203-41q109 0 202 41t163 112 108 166 39 203v374h256zm-1152 0h768V522q0-81-29-152t-80-126-122-85-153-31q-82 0-152 31t-122 85-81 125-29 153v374zm1024 128H384v896h1280v-896z"
                  />
                </svg>
              </span>
            </span>
          </div>
          <div
            class="emotion-4"
          >
            <span
              class="ui-text"
              data-uses-unhanded-props="true"
              dir="auto"
            >
              Welcome to the HR Taskforce team.
            </span>
          </div>
          <div
            class="emotion-8"
          >
            <div
              data-testid="ChannelMenuItem"
            />
            <div
              data-testid="MemberMenuItem"
            />
            <div
              data-testid="DriveMenuItem"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`TeamCard create shapshot when item is loading 1`] = `
.emotion-0 {
  height: 6rem;
}

.emotion-1 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: auto 1fr auto;
  gap: 0.5rem;
}

.emotion-2 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: auto;
  gap: 0.25rem;
}

<body>
  <div>
    <div
      class="ui-card emotion-0"
      data-aa-class="Card"
      data-uses-unhanded-props="true"
      role="listitem"
    >
      <div
        class="ui-skeleton"
        data-uses-unhanded-props="true"
      >
        <div
          class="emotion-1"
        >
          <span
            class="ui-skeleton__avatar"
            data-uses-unhanded-props="true"
          />
          <div
            class="emotion-2"
          >
            <span
              class="ui-skeleton__line"
              data-uses-unhanded-props="true"
            />
            <span
              class="ui-skeleton__line"
              data-uses-unhanded-props="true"
            />
            <span
              class="ui-skeleton__line"
              data-uses-unhanded-props="true"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
